tycho-extras/tycho-pomless/src/main/java/org/eclipse/tycho/pomless/AbstractXMLTychoMapping.java

```diff
@@ -96,4 +98,11 @@ protected static String getXMLAttributeValue(Element element, String attributeNa
     public float getPriority() {
         return 1;
     }
+
+    protected static Stream<File> filesWithExtension(File projectRoot, String extensions) throws IOException {
+        Predicate<String> nameFilter = n -> !n.startsWith(".polyglot.") && n.endsWith(extensions);
+        return Files.walk(projectRoot.toPath(), 1) //
```

Comment:
In the past I have seen problem with some obscure (windows) directories that could not be queried and listFiles() returning null, I wonder if `projectRoot.toPath()` is guaranteed to always return non `null`?

extensions --> extension ?
projectRoot --> directory?

Also one could think about passing a Path object instead of a File object here, and even though we need to have more file.toPath calls at the caller side, we might later on use more path API without converting back and forth.

Reply:
> In the past I have seen problem with some obscure (windows) directories that could not be queried and listFiles() returning null, I wonder if `projectRoot.toPath()` is guaranteed to always return non `null`?

I assumed that. But now I checked the doc and nothing of `File.toPath()` and `FileSystem.getPath(String)` (which does the work), mentions any case where null is returned. Unfortunately my Eclipse does not show me any implementation of FileSystem so I cannot verify it at the moment.

> 
> extensions --> extension ? projectRoot --> directory?

Done, thanks.

> 
> Also one could think about passing a Path object instead of a File object here, and even though we need to have more file.toPath calls at the caller side, we might later on use more path API without converting back and forth.

Sure. Actually personally I prefer the Path API over the the File API, because many thinks are just more convenient and easier to make performant. However I didn't want to change everything just for my personal preference. And unfortunately the Maven API also uses File. But I can change the parameter. However changing the return value requires so many back-conversion at the callers that I think it is not profitable at the moment.

Reply:
If its not in the API we don't need to handle it but [File.listFiles](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles()) says:

> Returns null if this abstract pathname does not denote a directory, or if an I/O error occurs.

I thin the "if an I/O error occurs" is covered here by throwing IOException already....

Reply:
> I thin the "if an I/O error occurs" is covered here by throwing IOException already....

Yes and furthermore `.toPath()` should not perform actual file-system operations. IIRC it only throws if for example an invalid file-system is used or the path cannot be parsed. But then one usually cannot create a File object either.